[MIG][16.0] stock_picking_group_by_partner_by_carrier

rousseldenis opened the pull request:
Depends on :

- https://github.com/odoo/odoo/pull/108452

gurneyalex committed:
[ADD] stock_picking_group_by_partner_by_carrier

This module can be used if your customers expect that several different orders
they passed will be shipped in a single delivery order.

With this module installed, when a sale order is confirmed, the stock moves for
the lines of the sale order can be placed in an existing delivery order that
shares the same delivery address and carrier (or lack thereof).

Sale orders with a Shipping Policy set to 'When all products are ready' always
get their own shipping.

gurneyalex committed:
[FIX] cancellation of sale order and WH multi step

* the grouped picking must not be cancelled

* in WH multi step mode:
** ensure the picking ids are computed correctly
** the sale order cancelation does what is expected

sebalix committed:
[FIX] group pick operations as well with pick-ship delivery steps

sebalix committed:
[FIX] stock_picking_group_by_partner_by_carrier: add a hook to handle domain on move_type

And add a test with the 'pick_pack_ship' configuration with grouping
enabled on pack+ship (but not pick).

OCA-git-bot committed:
[ADD] icon.png

TDu committed:
[13.0][IMP] stock_picking_group_by_partner_by_carrier

Improve the generated delivery note report. By adding the customer reference
of the sale order on each section.

grindtildeath committed:
stock_picking_group_by_partner_by_carrier: Display carrier on PG

Commit:
Fix groups of transfers on sales

When grouping transfers of different sales orders in the same
procurement.group, we have an issue with only the first sales order
showing the "Transfers" button on the SO view, because
"sale.picking_ids" returns an empty recordset on the other sales orders.

This is because the link between sales and pickings is done through the
"group_id.sale_id" field on the stock moves. When we group transfers of
different sales, we end up with several group_id in the moves of the
same picking, which is a bit unexpected, and when new transfers are
created (dynamic routing, backorders, ...), they'll only get the first
group_id and be attached only to the first sale order.

The solution implemented here is the following:

* Add a M2m "sale_ids" field on the procurement group
* When several moves with different groups are merged into the same
  group, clone the group, update its "sale_ids" field to contain all the
  sales (sale_id will keep the first one though) and set in on the moves
* Update the computed fields for the relation between the sales and
  pickings to use this M2M field
* Keep the original group in the stock move as "original_group_id": used
  when a sales order is cancelled

Commit:
Simplify line group comparison

Commit:
Prevent moves being always merged in the same groups

If we have a unique transfer for 2 sales orders, we have one joint
procurement group for them. If we delivery the first sale order, we
expect to no longer have the first SO in the group of the backorder.

This will prevent to have a never-ending group with all the orders from
the beginning if we have many backorders.

To so do, we consider that a "printed" or done transfer will no longer
change and therefore, we consider its group frozen.

Commit:
Rename/move around methods for clarity

TDu committed:
[13.0][IMP] stock_picking_group_by_partner_by_carrier

The delivery slip can reference more than one sale order with different
customer references. If there is more than one of those references they
will be displayed in the lines.
But only one in the header, better not to display the customer reference
in the header with multiple sale order.
Also fix the customer reference to be specific to the delivery lines.

jbaudoux committed:
Fix cancelation of SO when multiple SO are grouped in a same delivery.

Propagate the original_group_id on pull moves to allow to prevent
merging pulled moves. This allows to cancel a SO without canceling
pulled moves from other SO, as we ensure they are not merged.

squash! Fix cancelation of SO when multiple SO are grouped in a same delivery.

Add tests

AaronHForgeFlow committed:
[FIX]stock_picking_group_by_partner_by_carrier
* add valid development status

TDu committed:
[13.0][IMP] stock_picking_group_by_partner_by_carrier

Remove the sentence about remaining items to be shipped.
As it is confusing with multiple sale order in a delivery

simahawk committed:
stock_picking_group_by_* rebase on stock_move_assign_picking_hook

simahawk committed:
s_picking_group_by_partner_by_carrier: add merge wizard

sebalix committed:
s_picking_group_by_partner_by_carrier: add a regression test when creating backorder

If the group pickings option is disabled, the creation of a backorder
when validating a partial qty of a picking should work as usual.

sebalix committed:
s_picking_group_by_partner_by_carrier: fix regression when creating backorder

TDu committed:
[13.0][FIX] stock_picking_group_by_partner_by_carrier

When a sale order is cancelled then set to quotation again, all
related picking are cancelled also.

If before reconfirming the sale order for a 2nd time the delivery
carrier is changed, the procurement group linked on the sale order
will be used and related to the wrong carrier.

Removing the assigned procurement group on the reactivation
of the sale order fixes the issue.

TDu committed:
s_picking_group_by_partner_by_carrier: Fix MockedMove

Using an alternative data structure to store the separation of the
sale order on the delivery note as been sensible to breakage from
changes upstream.

Using the same type of records than the rest of the information
will hopefully help.

phuctranfxvn committed:
[IMP] stock_picking_group_by_partner_by_carrier: black, isort, prettier

Commit:
[IMP] stock_picking_group_by_partner_by_carrier

When the delivery slip is printed, the list of pending quantities
to deliver is shown at the end, grouped by order.

phuctranfxvn committed:
[MIG] stock_picking_group_by_partner_by_carrier: Migration to 14.0

sebalix committed:
s_picking_group_by_*: disable group on partner option

Allow to disable the grouping feature on partner and fallback on the
standard Odoo behavior.

sebalix committed:
[FIX] s_p_g_by_partner_by_carrier: remaining to deliver

Several moves could be related to a sale order line, qty should be
computed from all of them.

+ regarding the delivered qties, print only the moves having some
qty reserved if the transfer is not yet done.

mmequignon committed:
[13.0][IMP] s_p_g_by_partner_by_carrier: Add hook on delivery report

mmequignon committed:
s_p_group_by_partner_by_carrier: Fix translations in report

sebalix committed:
[IMP] s_p_g_by_p_by_carrier: hook to check state

oca-travis committed:
[UPD] Update stock_picking_group_by_partner_by_carrier.pot

OCA-git-bot committed:
[UPD] README.rst

sebalix committed:
[IMP] s_p_group_by_partner_by_carrier: black, isort, prettier

sebalix committed:
[FIX] s_p_group_by_partner_by_carrier: fix merge wizard

oca-travis committed:
[UPD] Update stock_picking_group_by_partner_by_carrier.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.1.0

jbaudoux committed:
[FIX] Properly merge the procurement groups

In case of a inter-WH resupply delivery order (i.e. not linked to any
SO), the grouping was effective but the procurement group was not merged
properly.

In case of SO, when adding new lines, the procurement group on the new
line was not correct.

In multi-step delivery, the new procurement group was not propagated all
to the new internal moves.

Do not create a new procurement group when not necessary and a rename
can be applied.

Sort procurement group name and origin by name.

Post message on picking that it has been modified when new moves from SO
are added (instead of re-posting that it has been created).

Also update origin in case of inter-WH resupply.

sebalix committed:
[FIX] s_p_group_by_: fix printing of delivery slip

Two bugfixes:
- filter out canceled moves
- in case there is no sale order related to some moves (even if this should
  not happen), do not print an empty section line

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.1.1

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.1.2

OCA-git-bot committed:
[UPD] README.rst

TDu committed:
Add hooks on stock_picking_group_by_partner_by_carrier

This adds two hooks to enable on the report:

* An alternative qantity calculation
* Easy customization of sales data

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.2.0

jbaudoux committed:
Add procurement_group_carrier. Extract feature from stock_picking_group_by_partner_by_carrier

simahawk committed:
s_picking_group_by_partner_by_carrier: dev status -> Beta

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.3.0

oca-travis committed:
[UPD] Update stock_picking_group_by_partner_by_carrier.pot

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.4.0

mmequignon committed:
Added translation using Weblate (German)

mmequignon committed:
Translated using Weblate (German)

Currently translated at 1.9% (1 of 52 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_group_by_partner_by_carrier/de/

mmequignon committed:
Added translation using Weblate (French)

mmequignon committed:
Translated using Weblate (French)

Currently translated at 9.6% (5 of 52 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_group_by_partner_by_carrier/fr/

mmequignon committed:
Added translation using Weblate (Italian)

mmequignon committed:
Translated using Weblate (Italian)

Currently translated at 1.9% (1 of 52 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_group_by_partner_by_carrier/it/

mmequignon committed:
s_p_group_by_partner_by_carrier: Fix deliveryslip report

https://github.com/odoo/odoo/blob/ef0320b82e679f2f6eb4b2ba8aa4530cda9a98a9/addons/stock/report/report_deliveryslip.xml#L77-L124

There's 4 different cases for the deliveryslip report:
 1) package and serial number
 2) package and no serial number
 3) no package but serial number
 4) no package and no serial number

For the 2nd an 4th case, odoo tries to aggregate the lines and therefore
shouldn't be groupped by SO, no fix needed here.

However, when there's serial numbers, we can group lines by SO, and we
should use `picking.get_delivery_report_lines()` to get the lines to
display. The 3rd case wasn't handled yet.

Also, when groupping by SO, the `get_delivery_report_lines` creates fake
`stock.move.line` for each SO section, with product_id = 1.
However, odoo tries to display the product, which is wrong for the
SO sections.

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.4.1

OCA-git-bot committed:
stock_picking_group_by_partner_by_carrier 14.0.1.4.2

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 5.7% (3 of 52 strings)

Translation: stock-logistics-workflow-14.0/stock-logistics-workflow-14.0-stock_picking_group_by_partner_by_carrier
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-14-0/stock-logistics-workflow-14-0-stock_picking_group_by_partner_by_carrier/it/

rousseldenis committed:
[14.0][FIX] Align development statuses

OCA-git-bot committed:
[UPD] README.rst

rousseldenis committed:
[MIG][16.0] stock_picking_group_by_partner_by_carrier

rousseldenis commented:
/ocabot migration stock_picking_group_by_partner_by_carrier

rousseldenis commented on stock_picking_group_by_partner_by_carrier/report/report_delivery_slip.xml:
```diff
@@ -0,0 +1,176 @@
+
+        <xpath
+            expr='//table[@name="stock_move_table"]/tbody/tr//span[@t-field="move.product_id"]/..'
+            position="replace"
```
@TDU @jbaudoux I don't like replaces but I haven't found a better method at this time

simahawk replied:
hide it setting attribute `t-if=False` then add your later

rousseldenis commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -0,0 +1,114 @@
+
+        return domain + grouping_domain
+
+    # TODO: this part and everything related to generic grouping
```
@simahawk @jbaudoux What do you think of this ?

How choosing the good domain as I suppose several group_by_... modules can be installed together

simahawk replied:
Yeah, you can have more than one. So far, with the ones available today we didn't have problems.
If you need to take full control, you should probably override this in your custom module.
The idea here is to have a base module for defining an hook and some sort of api for this feat.

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: printed field is copy=False in core

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: Use attirbutes instead of replace

rousseldenis committed:
[IMP] stock_picking_group_by_partner_by_carrier: Add index on grouping fields

lmignon commented:
@rousseldenis 
Also depends on
 * [x] #1158
 * [x] #1187
> Depends on :
> 
>     * [[FIX] delivery: Really use propagate_carrier option on stock rules odoo/odoo#108452](https://github.com/odoo/odoo/pull/108452)

lmignon reviewed (approved):
Thank you @rousseldenis LGTM (Code review + functional tests)

FrancoMaxime reviewed (approved):
code review

sebalix commented:
@rousseldenis the dependency has been merged, you can drop your `[DONT MERGE]` commit

glitchov committed:
[IMP] stock_picking_group_by_partner_by_carrier : fr translations

jbaudoux reviewed (approved):
@rousseldenis Can you squash the "DO NOT MERGE" commits?

jbaudoux commented:
@rousseldenis @glitchov Can you drop the "DO NOT MERGE" commit and its revert ? Then we can merge this PR

hparfr commented:
@rousseldenis is it ready for  merge ?

rousseldenis commented:
> @rousseldenis is it ready for merge ?

Of course

rousseldenis commented:
@simahawk Any more comment ?

simahawk commented on stock_quant_package_dimension/README.rst:
```diff
@@ -7,9 +7,9 @@ Stock Quant Package Dimension
    !! changes will be overwritten.                   !!
```
The change in [stock_quant_package_dimension](https://github.com/OCA/stock-logistics-workflow/pull/1165/files#diff-5cc932375429134f87e8cf556c589feeb173fa168716649e0c63036f08305fe4) seems not relevant.

simahawk reviewed (approved):
LG

There's a typo in the commit msg "Use attirbutes instead..." s/attirbutes/attributes.
Also the dev state has been changed in that commit... was it by decision?

lmignon committed:
[IMP] stock_picking_group_by_partner_by_carrier: Add index on picking.group_id

In the '_merge_procurement_groups' from the 'stock.picking' model, pickings are browsed from the group_id field of a stock_move ('group_pickings = self.move_ids.group_id.picking_ids.filtered'. We add an index on the 'group_id' field on the 'stock_picking' model to make this process efficient on large database

rousseldenis commented:
/ocabot merge nobump
